v1.3: Experimental features API route (#2496)
meilisearch/documentation: learn/experimental/overview.mdx

# Experimental features

> Meilisearch periodically introduces new experimental features. Experimental features are not yet ready for production, but offer functionality that might benefit some users.

Author: Would you say the same about `--experimental-reduce-indexing-memory-usage`, @dureuill? My understanding was that this feature was tagged as experimental not because we're unsure about stabilising it, but because we'd like to see it in action before integrating it in our codebase.

> An experimental feature's API can change significantly and become incompatible between releases. Meilisearch does not recommend using experimental features in a production environment.

Author: Huh, I guess this was recently updated? Last info I had received was that we did not encourage using experimental features in production.

Here's a lighter version. If you still think it's too strong, I'll remove the second sentence altogether.

```suggestion
An experimental feature's API can change significantly and become incompatible between releases. Keep this in mind when using experimental features in a production environment.
```

Meilisearch makes experimental features available expecting they will become stable in a future release. However, it is not possible to guarantee when and if this will happen.

## Configuring experimental features

There are two main ways to activate and deactivate experimental features. Some must be configured at launch with a command-line flag or an environment variable. Others can only be configured with the `/experimental-features` API route.

## Configuring experimental features at launch

Launch Meilisearch using [configuration options](/learn/configuration/instance_options) to activate an experimental feature:

```sh
./meilisearch --experimental-enable-metrics
```

Consult the documentation for the specific features you are interested in to find their respective options. Command-line flags for experimental features are always prefixed with `--experimental`. Environment variables for experimental features are always prefixed with `MEILI_EXPERIMENTAL`.

Activating or deactivating experimental features this way requires you to relaunch Meilisearch.

> ### Configuring experimental features during runtime

Reviewer: ```suggestion
### Activating experimental features during runtime
```

> [Use the `/experimental-features` API route](/reference/api/experimental) to configure experimental features during runtime:

Reviewer: ```suggestion
Some experimental features can be activated via an HTTP call using the [`/experimental-features` API route](/reference/api/experimental):
```

<CodeSamples id="update_experimental_features_1" />

Activating or deactivating experimental features this way does not require you to relaunch Meilisearch.

## Current experimental features

| Name                                                                             | Description                                                | How to configure                                  |
|----------------------------------------------------------------------------------|------------------------------------------------------------|---------------------------------------------------|
| [Metrics API](/learn/experimental/metrics)                                       | Exposes Prometheus-compatible analytics data               | At launch with a CLI flag or environment variable |
| [Ranking score details](/learn/experimental/ranking_rule_score_details)          | Exposes advanced result ranking score data                 | During runtime with the API route                 |
| [Reduce indexing memory usage](/learn/experimental/reduce-indexing-memory-usage) | Optimizes indexing performance                             | At launch with a CLI flag or environment variable |

> | [Vector search](/learn/experimental/vector_search)                               | Allows Meilisearch to function as a vector embedding store | During runtime with the API route                 |

Reviewer: Seems like you could simplify the value of the last column of the table: either "CLI flag or env variable" or "`/experimental-features` API route"

I would caution against using this phrase "during runtime." I think the word runtime is more often used in the set phrase "at runtime", which is itself ambiguous as it can mean at launch or at any point in the program's execution.

In general, I also think this concept of "runtime" is more common with programs that follow a fairly set process of execution, rather than server-based APIs that listen for calls. TL;DR I would just avoid the phrase "during runtime" if possible since it is ambiguous.

# Experimental features

> Meilisearch periodically introduces new experimental features. Experimental features are not always ready for production, but offer functionality that might benefit some users.

Author: I was discussing this with @dureuill in a separate conversation. In short, while it is true some experimental features are mostly ready and only marked as experimental because of possible API changes (e.g. metrics), that is not the case for others. For example, my understanding is that `--experimental-reduce-indexing-memory-usage` is experimental not because its API might change, but because we want to see it in the wild before fully integrating it into our codebase. @dureuill mentioned vector search implementation is also not as strong and thorough as a "regular" feature.

Prior to @dureuill's comment, this sentence read `Experimental features are not ready for production, […]`. We had more or less agreed on qualifying this statement by mentioning not all experimental features are ready for production.

If you still think it's problematic, I could add an extra sentence advising users to check a feature's dedicated page/discussion for more information.

I do think we need to keep some sort of warning, though. I understand the concern it might scare users away, but it seems misleading to say "experimental features are only experimental because the API may change" when that only applies to some of them.
